File: camelcase_matching/Camelcase_Matching.cpp
class Camelcase_Matching {
public:
    vector<bool> camelMatch(vector<string>& queries, string pattern) {
        if (queries.empty()) {
            return vector<bool>();
        }
        if (pattern.empty()) {
            return vector<bool>(queries.size(), false);
        }

        vector<bool> camel_matches;
        for (string& query : queries) {
            camel_matches.push_back(camel_match(query, pattern));
        }
        return camel_matches;
    }

private:
    bool camel_match(string& query, string& pattern) {
        int ptr_p = 0;
        int ptr_q = 0;
        while (ptr_q < query.size() && ptr_p < pattern.size()) {
            if (query[ptr_q] == pattern[ptr_p]) {
                ptr_p++;
                ptr_q++;
                continue;
            }
            if (isupper(query[ptr_q++])) {
                return false;
            }
        }
        while (ptr_q < query.size() && islower(query[ptr_q])) {
            ptr_q++;
        }

        return ptr_p == pattern.size() && ptr_q == query.size();
    }
};

File: camelcase_matching/CamelcaseMatching.java
class CamelcaseMatching {
    public List<Boolean> camelMatch(String[] queries, String pattern) {
        if (queries == null || queries.length == 0) {
            return new ArrayList<Boolean>();
        }

        List<Boolean> camelMatches = new ArrayList<>();
        if (pattern == null || pattern.length() == 0) {
            Collections.fill(camelMatches, false);
            return camelMatches;
        }

        for (String query : queries) {
            camelMatches.add(camelMatch(query, pattern));
        }

        return camelMatches;
    }

    private boolean camelMatch(String query, String pattern) {
        int qPtr = 0;
        int pPtr = 0;
        while (qPtr < query.length() && pPtr < pattern.length()) {
            char q = query.charAt(qPtr++);
            char p = pattern.charAt(pPtr);
            if (p == q) {
                pPtr++;
                continue;
            }
            if (Character.isUpperCase(q)) {
                return false;
            }
        }
        while (qPtr < query.length() && Character.isLowerCase(query.charAt(qPtr))) {
            qPtr++;
        }

        return qPtr == query.length() && pPtr == pattern.length();
    }
}